packages/caravan-fees/src/rbf.ts

```diff
@@ -0,0 +1,530 @@
+      const input = availableInputs.find(
+        (utxo) => utxo.txid === template.txid && utxo.vout === template.vout,
+      );
+      if (input) {
+        newTxTemplate.addInput(BtcTxInputTemplate.fromUTXO(input));
+        addedInputs.add(`${input.txid}:${input.vout}`);
+      }
+    });
+  } else {
+    // Add at least one input from the original transaction
+
+    // TO DO (MRIGESH)
+    // Add coin selection algorithm to add best suited input to start with ...
+    const originalInput = availableInputs.find((utxo) =>
+      originalInputTemplates.some(
+        (template) =>
+          template.txid === utxo.txid && template.vout === template.vout,
+      ),
+    );
+    if (!originalInput) {
+      throw new Error(
+        "None of the original transaction inputs found in available UTXOs.",
+      );
+    }
+    newTxTemplate.addInput(BtcTxInputTemplate.fromUTXO(originalInput));
+    addedInputs.add(`${originalInput.txid}:${originalInput.vout}`);
+  }
+
+  // Step 6: Add more inputs if necessary to meet fee requirements
+
+  // We continue adding inputs until both the fee rate is satisfied and
+  // the absolute fee meets the minimum RBF requirement
+  for (const utxo of availableInputs) {
+    if (
+      newTxTemplate.feeRateSatisfied &&
+      new BigNumber(newTxTemplate.currentFee).gte(
+        BigNumber.max(newTxTemplate.targetFeesToPay, originalTxFee),
+      )
+    ) {
+      // Stop adding inputs when both conditions are met:
+      // 1. The fee rate satisfies the target rate
+      // 2. The current fee is greater than or equal to the maximum of:
+      //    a) The target fee based on the desired rate by user
+      //    b) The minimum fee required for a valid RBF replacement >= original fees
+      break;
+    }
+    // Skip if this UTXO is already added
+    const utxoKey = `${utxo.txid}:${utxo.vout}`;
+    if (addedInputs.has(utxoKey)) {
+      continue;
+    }
+    newTxTemplate.addInput(BtcTxInputTemplate.fromUTXO(utxo));
+    addedInputs.add(utxoKey);
+  }
+
+  // Step 7: Calculate and set the cancellation output amount
+
+  const totalInputAmount = new BigNumber(newTxTemplate.getTotalInputAmount());
+  // Ensure we're using the higher of the target fee and the original fee
+  const fee = BigNumber.max(newTxTemplate.targetFeesToPay, originalTxFee);
+  // The cancel output receives all funds minus the fee
+  const cancelOutputAmount = totalInputAmount.minus(fee);
+
+  if (cancelOutputAmount.isLessThan(dustThreshold)) {
+    if (strict) {
+      throw new Error(
+        "Cancel output would be dust. Increase inputs or fee rate.",
+      );
+    } else {
+      console.warn(
+        "Cancel output amount is below dust threshold. Adjusting transaction.",
+      );
+      newTxTemplate.outputs[0].setAmount("0");
+    }
+  } else {
+    newTxTemplate.outputs[0].setAmount(cancelOutputAmount.toString());
+  }
+
+  // Step 8: Ensure the new transaction meets RBF requirements
```

Comment:
this could be its own function as well that's used in both.

Reply:
A lot of this could probably be simplified too. This should throw if:

1. new template fee is less than analyzer min RBF fee
2. target fees are not paid
3. the transaction creates dust
4. the template doesn't validate.

Reply:
changes [done](https://github.com/caravan-bitcoin/caravan/pull/114/commits/c34de8a8ff6115cd401f7d5cb16406c991140064#diff-47122bb461c82a41c84d8866adf0cb005bf7f573a2afb8558f670a419bed146aR322)